### PR TITLE
Magazine Viewer, Magazine by Publications, Bookmark Magazines

### PR DESCRIPTION
**Pull request by zachseidner1**
## Overview

In this PR, the individual magazine viewer, the ability to view magazines by publication, and the ability to bookmark and view bookmarked magazines has been added, thus completing the magazines feature for Android. 

## Changes Made

- Implemented a PDF viewer for Jetpack compose with [Bouquet](https://github.com/GRizzi91/bouquet)
- Implemented loading bar for PDF download progress in individual viewer
- Magazine covers now load the first page of the magazine with mini-PDF viewers.
- Magazines can be bookmarked, shared, and shouted out. 
- Bookmarks page now has a section for viewing bookmarked magazines.
- Refactored publications page to use a lazy vertical grid to support both article and magazine viewing. 
- Change default selection from semester view to a "View all" button, which shows all available magazines. 

## Please note:

- Some features seem to only work on the prod server, such as loading featured magazines or loading all magazines. This _should_ not be an issue when the app is shipped to the Play Store.
- Following publications can be a bit inconsistent depending on what server you are on, again this should not be an issue when the app is shipped.
- On emulators, the PDF viewer can be a bit laggy. I implemented everything I could using the tools available in the library to optimize the experience. On the Android test phone, this behavior is **not** shown.

## Next Steps

Magazines are finished! Yay me. Next step will be implementing the community board, and potentially adding some smaller features and bug fixes to bring feature parity. Also will probably be implementing dedicated cover images instead of loading a lot of mini-PDF viewers which will optimize the experience.

## Screenshots & Videos

<details>
  <summary>Viewing a Magazine with the PDF Viewer</summary>
 
</details>

<details>
  <summary>Viewing Bookmarked Magazines</summary>
 
</details>

<details>
  <summary>Viewing Magazines By Publication</summary>
 
</details>

### CODE REVIEW

**Comment by Corfish123** [comment]
I tried running the app and there were some issues I noted. First, when I am in the magazine tab, and I am under View All, it is constantly loading and nothing ever shows up. From the log statements it seems like you don't actually do any calling as this was the network call: {"data":{"getMagazinesBySemester":[]}}. As a result nothing is loading

2. When you are on the magazine screen, and try scrolling, you are able to scroll, however, it is only possible if you are scrolling when your first touch is between the preview of the magazine. If you try to scroll when you finger is on the preview of the magazine, it does not scroll. They are able to scroll while they are still loading with the shimmer effect but once they load in, you can no longer press there to scroll. As a result, this causes a lot of dead zones where you can not scroll. It seems the issue might be the pdf previewer composable that appears. 
image
So basically all the highlighted areas you can scroll from there. 

3. You also mentioned that you made tabs for the publications here
image
However, I do not see the tab appear in this PR.

File: app/src/main/java/com/cornellappdev/android/volume/ui/screens/HomeScreen.kt [hunk 1]
```diff
@@ -87,36 +88,55 @@ fun HomeScreen(
...
+                                    onArticleClick(
+                                        article,
+                                        NavigationSource.FOLLOWING_ARTICLES
+                                    )
                                 }
-                                showPageBreak = true
                             }
                         }
+                        showPageBreak = true
+//                            Box(modifier = Modifier.padding(top = 10.dp)) {
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/android/volume/ui/screens/HomeScreen.kt]: Delete this when pushing the main

File: app/src/main/java/com/cornellappdev/android/volume/ui/screens/IndividualMagazineScreen.kt [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+import com.cornellappdev.android.volume.ui.theme.VolumeOrange
+import com.cornellappdev.android.volume.ui.theme.lato
+import com.cornellappdev.android.volume.ui.theme.notoserif
+import com.cornellappdev.android.volume.ui.viewmodels.IndividualMagazineViewModel
+import com.rizzi.bouquet.HorizontalPDFReader
+import com.rizzi.bouquet.ResourceType
+import com.rizzi.bouquet.rememberHorizontalPdfReaderState
+
+private const val TAG = "IndividualMagazineScreen"
+
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/android/volume/ui/screens/IndividualMagazineScreen.kt]: Add a comment here

File: app/src/main/java/com/cornellappdev/android/volume/ui/screens/IndividualPublicationScreen.kt [hunk 3]
```diff
@@ -1,46 +1,55 @@
...
+        item (span = { GridItemSpan(2)}){
             when (val publicationState = publicationUiState.publicationState) {
                 PublicationRetrievalState.Loading -> {
-                    Column(
-                        modifier = Modifier.fillMaxWidth(),
-                        horizontalAlignment = Alignment.CenterHorizontally
-                    ) {
-                        CircularProgressIndicator(color = VolumeOrange)
-                    }
+                    Log.d(TAG, "IndividualPublicationScreen: Loading state reached")
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/android/volume/ui/screens/IndividualPublicationScreen.kt]: delete these log statements

File: app/src/main/java/com/cornellappdev/android/volume/ui/screens/MagazinesScreen.kt [hunk 4]
```diff
@@ -166,31 +179,38 @@ fun MagazinesScreen(
...
+
+
+@RequiresApi(Build.VERSION_CODES.P)
 @Composable
-fun FillFeaturedMagazinesRow(magazineUiState: MagazinesViewModel.MagazinesUiState) {
+fun FillFeaturedMagazinesRow(magazineUiState: MagazinesViewModel.MagazinesUiState, onMagazineClick: (magazine: Magazine) -> Unit)
+{
     when (val magazinesState =  magazineUiState.featuredMagazinesState) {
         MagazinesRetrievalState.Loading -> {
+            Log.d(TAG, "FillFeaturedMagazinesRow: MAGAZINE LOADING")
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/android/volume/ui/screens/MagazinesScreen.kt]: delete these logs as well

File: app/src/main/java/com/cornellappdev/android/volume/ui/viewmodels/IndividualPublicationViewModel.kt [hunk 5]
```diff
@@ -59,18 +60,22 @@ class IndividualPublicationViewModel @Inject constructor(
 
     private fun queryPublication() = viewModelScope.launch {
         try {
+            Log.d(TAG, "queryPublication: No null pointers")
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/android/volume/ui/viewmodels/IndividualPublicationViewModel.kt]: delete these log statements

File: app/src/main/java/com/cornellappdev/android/volume/ui/viewmodels/IndividualPublicationViewModel.kt [hunk 6]
```diff
@@ -79,16 +84,34 @@ class IndividualPublicationViewModel @Inject constructor(
...
-            publicationUiState.copy(
+             publicationUiState = publicationUiState.copy(
                 articlesByPublicationState = ArticlesRetrievalState.Error
             )
         }
+        queryMagazinesByPublication()
+    }
+
+    private fun queryMagazinesByPublication() = viewModelScope.launch {
+        Log.d(TAG, "queryMagazinesByPublication: Querying magazines")
```
Comment (Corfish123) [inline app/src/main/java/com/cornellappdev/android/volume/ui/viewmodels/IndividualPublicationViewModel.kt]: delete this

**Review by Corfish123 — COMMENTED** [review]
Keep up the good work!! Just remember to delete all of your testing log statements before pushing it

**Review by Corfish123 — APPROVED** [review]
Good stuff!! Keep up the good work
